src/rust/ide/lib/span-tree/src/node/kind.rs

```diff
@@ -0,0 +1,303 @@
+}
+
+
+// === Smart Setters ===
+
+#[allow(missing_docs)]
+impl Argument {
+    pub fn removable      (mut self)                      -> Self { self.removable = true  ; self }
+    pub fn named          (mut self, name:String)         -> Self { self.name = Some(name) ; self }
+    pub fn typed          (mut self, tp:String)           -> Self { self.tp = Some(tp)     ; self }
+    pub fn with_removable (mut self, rm:bool)             -> Self { self.removable = rm    ; self }
+    pub fn with_name      (mut self, name:Option<String>) -> Self { self.name = name       ; self }
+    pub fn with_tp        (mut self, tp:Option<String>)   -> Self { self.tp = tp           ; self }
+}
+
+impl From<Argument> for Kind {
+    fn from(t:Argument) -> Self {
+        Self::Argument(t)
+    }
+}
+
+
+
+// ======================
+// === InsertionPoint ===
+// ======================
+
+/// Kind representing "insertion point" node. For example, in the following expressions, all places
+/// marked with `?` are considered "insertion points" (please note that `?` symbols are NOT part
+/// of the source code): `foo ? a ? b ? c ? ? ?`. The final question marks indicate non-provided
+/// arguments.
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub struct InsertionPoint {
+    pub kind : InsertionPointType,
+    pub name : Option<String>,
+    pub tp   : Option<String>,
+}
+
+// === Smart Constructors ===
+
+#[allow(missing_docs)]
+impl InsertionPoint {
+    pub fn before_target() -> Self { Self::default().with_kind(InsertionPointType::BeforeTarget) }
+    pub fn after_target()  -> Self { Self::default().with_kind(InsertionPointType::AfterTarget) }
+    pub fn append()        -> Self { Self::default().with_kind(InsertionPointType::Append) }
+    pub fn expected_argument(ix:usize) -> Self {
+        Self::default().with_kind(InsertionPointType::ExpectedArgument(ix))
+    }
+}
+
+
+// === Smart Setters ===
+
+#[allow(missing_docs)]
+impl InsertionPoint {
+    pub fn named     (mut self, name:String)             -> Self { self.name = Some(name) ; self }
+    pub fn typed     (mut self, tp:String)               -> Self { self.tp   = Some(tp)   ; self }
+    pub fn with_kind (mut self, kind:InsertionPointType) -> Self { self.kind = kind       ; self }
+    pub fn with_name (mut self, name:Option<String>)     -> Self { self.name = name       ; self }
+    pub fn with_tp   (mut self, tp:Option<String>)       -> Self { self.tp   = tp         ; self }
+}
+
+impl From<InsertionPoint> for Kind {
+    fn from(t:InsertionPoint) -> Self {
+        Self::InsertionPoint(t)
+    }
+}
+
+
+
+// ==========================
+// === InsertionPointType ===
+// ==========================
+
+/// A helpful information about how the new AST should be inserted during Set action. See `action`
+/// module.
+#[allow(missing_docs)]
+#[derive(Copy,Clone,Debug,Eq,PartialEq)]
+pub enum InsertionPointType {
+    // FIXME: Why we need both before and after?
```

Comment:
This distinction is purely for operators: in expression a + b, you can put before a (BeforeTarget), after a (AfterTarget) or after b (Append) - all those InsertionPoints are on the same level, but their _set_ action should be handled differently.

Reply:
As agreed via phone, please add extensive documentation about these cases with nice ascii art tellign about it, as I would never understand what it is about without you drawing it to me today!